### PR TITLE
BUG: add reset logic for Grouper if new obj is passed in (#26564)

### LINKED ISSUE #26564: `pd.groupby` seems to mutate my `pd.Grouper` in-place
#### Code Sample (copy-pastable)

```python
import pandas as pd

df1 = pd.DataFrame([["a", 1, 2, "05/29/2019"], ["a", 4, 5, "05/28/2019"], ["b", 2, 3, "05/27/2019"]], columns=["type", "num1", "num2", "date"]).assign(date=lambda df: pd.to_datetime(df["date"]))
df2 = pd.DataFrame(columns=["type", "num1", "num2", "date"]).assign(date=lambda df: pd.to_datetime(df["date"]))

groupbys = ["type", pd.Grouper(key="date", freq="1D")]

df1.groupby(groupbys).head()
df2.groupby(groupbys).head()
```
#### Problem description

The above code sample throws the following error:

```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-ae6b527f8edf> in <module>
      7 
      8 df1.groupby(groupbys).head()
----> 9 df2.groupby(groupbys).head()

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/groupby.py in head(self, n)
   2062         """
   2063         self._reset_group_selection()
-> 2064         mask = self._cumcount_array() < n
   2065         return self._selected_obj[mask]
   2066 

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/groupby.py in _cumcount_array(self, ascending)
    730         (though the default is sort=True) for groupby in general
    731         """
--> 732         ids, _, ngroups = self.grouper.group_info
    733         sorter = get_group_index_sorter(ids, ngroups)
    734         ids, count = ids[sorter], len(ids)

pandas/_libs/properties.pyx in pandas._libs.properties.CachedProperty.__get__()

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/ops.py in group_info(self)
    250     @cache_readonly
    251     def group_info(self):
--> 252         comp_ids, obs_group_ids = self._get_compressed_labels()
    253 
    254         ngroups = len(obs_group_ids)

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/ops.py in _get_compressed_labels(self)
    266 
    267     def _get_compressed_labels(self):
--> 268         all_labels = [ping.labels for ping in self.groupings]
    269         if len(all_labels) > 1:
    270             group_index = get_group_index(all_labels, self.shape,

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/ops.py in <listcomp>(.0)
    266 
    267     def _get_compressed_labels(self):
--> 268         all_labels = [ping.labels for ping in self.groupings]
    269         if len(all_labels) > 1:
    270             group_index = get_group_index(all_labels, self.shape,

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/grouper.py in labels(self)
    365     def labels(self):
    366         if self._labels is None:
--> 367             self._make_labels()
    368         return self._labels
    369 

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/grouper.py in _make_labels(self)
    386             # we have a list of groupers
    387             if isinstance(self.grouper, BaseGrouper):
--> 388                 labels = self.grouper.label_info
    389                 uniques = self.grouper.result_index
    390             else:

pandas/_libs/properties.pyx in pandas._libs.properties.CachedProperty.__get__()

~/miniconda3/envs/mbc/lib/python3.6/site-packages/pandas/core/groupby/ops.py in label_info(self)
    261         labels, _, _ = self.group_info
    262         if self.indexer is not None:
--> 263             sorter = np.lexsort((labels, self.indexer))
    264             labels = labels[sorter]
    265         return labels

ValueError: all keys need to be the same shape
```

The `pd.Grouper` object seems to be modified inside the list. This is not the behavior I expect. I can resolve this by making explicit (deep) copies of the list so that new instances of `pd.Grouper` are passed into both `groupby` methods. Like so:

```python
In [11]: import copy                                                                                                                        

In [12]: groupbys = ["type", pd.Grouper(key="date", freq="1D")]                                                                             

In [13]: df1.groupby(copy.deepcopy(groupbys)).head()                                                                                        
Out[13]: 
  type  num1  num2       date
0    a     1     2 2019-05-29
1    a     4     5 2019-05-28
2    b     2     3 2019-05-27

In [14]: df2.groupby(copy.deepcopy(groupbys)).head()                                                                                        
Out[14]: 
Empty DataFrame
Columns: [type, num1, num2, date]
Index: []
```

#### Expected Output
As show above, expected output is the empty dataframe, instead of the `ValueError`. Interestingly, if I reverse the order and run the `df2.groupby` first, then run `df1.groupby`, it works fine. However, doing `df2.groupby` again throws the `ValueError`. There's definitely something in the `df1.groupby` that is modifying the `pd.Grouper`.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.7.final.0
python-bits: 64
OS: Linux
OS-release: 4.15.0-50-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 4.4.0
pip: 19.0.3
setuptools: 40.8.0
Cython: None
numpy: 1.16.2
scipy: 1.2.1
pyarrow: None
xarray: None
IPython: 7.4.0
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 3.0.3
openpyxl: None
xlrd: 1.2.0
xlwt: None
xlsxwriter: 1.1.6
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: 1.2.18
pymysql: None
psycopg2: 2.8.1 (dt dec pq3 ext lo64)
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

If this is expected behavior please let me know (we can close the issue). If this is not expected behavior, I'd love to take a crack at resolving this (any insight into the issue would be appreciated).

### PR DESCRIPTION
**Pull request by alichaudry**
- [ ] closes #26564
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -109,6 +109,13 @@ def __init__(self, key=None, level=None, freq=None, axis=0, sort=False):
         self.axis = axis
         self.sort = sort
 
+        self._reset_grouper()
+
+    def _reset_grouper(self):
+        """
+        Ensures Grouper object can be reused
+        See https://bit.ly/2D4Yd6V
```
Comment (jbrockmendel) [inline pandas/core/groupby/grouper.py]: can you use the real link

**Comment by jbrockmendel** [comment]
needs tests that the relevant bug is fixed

File: pandas/core/groupby/grouper.py (hunk 1 above)
Comment (alichaudry) [reply pandas/core/groupby/grouper.py]: yes, will change it shortly

**Comment by jreback** [comment]
this PR is likely duplicating this one: https://github.com/pandas-dev/pandas/pull/29131

**Comment by alichaudry** [comment]
> this PR is likely duplicating this one: #29131

Maybe I'll add my test here and see if it passes with #29131 .

File: pandas/tests/resample/test_resampler_grouper.py [hunk 2]
```diff
@@ -275,3 +275,23 @@ def test_median_duplicate_columns():
...
+def test_same_grouper_on_different_frames():
+
+    df1 = pd.DataFrame(
+        [
+            ["a", 1, 2, "05/29/2019"],
+            ["a", 4, 5, "05/28/2019"],
+            ["b", 2, 3, "05/27/2019"],
+        ],
+        columns=["type", "num1", "num2", "date"],
+    ).assign(date=lambda df: pd.to_datetime(df["date"]))
```
Comment (jbrockmendel) [inline pandas/tests/resample/test_resampler_grouper.py]: are the `assign` calls needed here?  if we aren't trying to test assign, itd be better to just write out df1 and df2 directly

File: pandas/tests/resample/test_resampler_grouper.py [hunk 3]
```diff
@@ -275,3 +275,23 @@ def test_median_duplicate_columns():
...
+    df2 = pd.DataFrame(columns=["type", "num1", "num2", "date"]).assign(
+        date=lambda df: pd.to_datetime(df["date"])
+    )
+
+    groupbys = ["type", pd.Grouper(key="date", freq="1D")]
+
+    df1.groupby(groupbys).head()
+    df2.groupby(groupbys).head()
```
Comment (jbrockmendel) [inline pandas/tests/resample/test_resampler_grouper.py]: are there functions other than `head` that are affected?

File: pandas/tests/resample/test_resampler_grouper.py (hunk 2 above)
Comment (alichaudry) [reply pandas/tests/resample/test_resampler_grouper.py]: no they're not needed. changes `df` construction to keep it simpler.

File: pandas/tests/resample/test_resampler_grouper.py (hunk 3 above)
Comment (alichaudry) [reply pandas/tests/resample/test_resampler_grouper.py]: yes, all aggregate functions as well. i was just using `head` as an example. I'll switch them to `sum` and `count` respectively.

**Comment by alichaudry** [comment]
@jreback I wrote a test [here](https://github.com/pandas-dev/pandas/pull/29800/files#diff-6f020a6c8d0f692c7f6e87cfe0775fe7R280) and ran it against #29131 which doesn't pass the test. So this is still an outstanding issue:

```
=========================================================== test session starts ============================================================
platform linux -- Python 3.7.3, pytest-5.3.0, py-1.8.0, pluggy-0.13.0
rootdir: /home/ali/repo/pydatascience/pandas-dev2, inifile: setup.cfg
plugins: cov-2.8.1, xdist-1.30.0, hypothesis-4.47.1, forked-1.1.2
collected 1 item                                                                                                                           

pandas/tests/resample/test_resampler_grouper.py F                                                                                    [100%]

================================================================= FAILURES =================================================================
__________________________________________________ test_same_grouper_on_different_frames ___________________________________________________

    def test_same_grouper_on_different_frames():
    
        df1 = pd.DataFrame(
            [["a", 1, 2], ["a", 4, 5], ["b", 2, 3]], columns=["type", "num1", "num2"],
        )
        df1["date"] = pd.to_datetime(["05/29/2019", "05/28/2019", "05/27/2019"])
    
        df2 = pd.DataFrame([["c", 6, 7], ["d", 8, 9]], columns=["type", "num1", "num2"],)
        df2["date"] = pd.to_datetime(["02/12/2018", "03/13/2018"])
    
        groupbys = ["type", pd.Grouper(key="date", freq="1D")]
    
        df1.groupby(groupbys).sum()
>       df2.groupby(groupbys).count()

pandas/tests/resample/test_resampler_grouper.py:294: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/core/groupby/generic.py:1674: in count
    ids, _, ngroups = self.grouper.group_info
pandas/_libs/properties.pyx:34: in pandas._libs.properties.CachedProperty.__get__
    val = self.func(obj)
pandas/core/groupby/ops.py:299: in group_info
    comp_ids, obs_group_ids = self._get_compressed_labels()
pandas/core/groupby/ops.py:315: in _get_compressed_labels
    all_labels = [ping.labels for ping in self.groupings]
pandas/core/groupby/ops.py:315: in <listcomp>
    all_labels = [ping.labels for ping in self.groupings]
pandas/core/groupby/grouper.py:405: in labels
    self._make_labels()
pandas/core/groupby/grouper.py:424: in _make_labels
    labels = self.grouper.label_info
pandas/_libs/properties.pyx:34: in pandas._libs.properties.CachedProperty.__get__
    val = self.func(obj)
pandas/core/groupby/ops.py:310: in label_info
    sorter = np.lexsort((labels, self.indexer))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = ((array([ 0, 29]), array([2, 1, 0])),), kwargs = {}, relevant_args = (array([ 0, 29]), array([2, 1, 0]))

>   ???
E   ValueError: all keys need to be the same shape

<__array_function__ internals>:6: ValueError
============================================================ 1 failed in 0.17s =============================================================
```

**Comment by jreback** [comment]
> @jreback I wrote a test [here](https://github.com/pandas-dev/pandas/pull/29800/files#diff-6f020a6c8d0f692c7f6e87cfe0775fe7R280) and ran it against #29131 which doesn't pass the test. So this is still an outstanding issue:
> 
> ```
> =========================================================== test session starts ============================================================
> platform linux -- Python 3.7.3, pytest-5.3.0, py-1.8.0, pluggy-0.13.0
> rootdir: /home/ali/repo/pydatascience/pandas-dev2, inifile: setup.cfg
> plugins: cov-2.8.1, xdist-1.30.0, hypothesis-4.47.1, forked-1.1.2
> collected 1 item                                                                                                                           
> 
> pandas/tests/resample/test_resampler_grouper.py F                                                                                    [100%]
> 
> ================================================================= FAILURES =================================================================
> __________________________________________________ test_same_grouper_on_different_frames ___________________________________________________
> 
>     def test_same_grouper_on_different_frames():
>     
>         df1 = pd.DataFrame(
>             [["a", 1, 2], ["a", 4, 5], ["b", 2, 3]], columns=["type", "num1", "num2"],
>         )
>         df1["date"] = pd.to_datetime(["05/29/2019", "05/28/2019", "05/27/2019"])
>     
>         df2 = pd.DataFrame([["c", 6, 7], ["d", 8, 9]], columns=["type", "num1", "num2"],)
>         df2["date"] = pd.to_datetime(["02/12/2018", "03/13/2018"])
>     
>         groupbys = ["type", pd.Grouper(key="date", freq="1D")]
>     
>         df1.groupby(groupbys).sum()
> >       df2.groupby(groupbys).count()
> 
> pandas/tests/resample/test_resampler_grouper.py:294: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> pandas/core/groupby/generic.py:1674: in count
>     ids, _, ngroups = self.grouper.group_info
> pandas/_libs/properties.pyx:34: in pandas._libs.properties.CachedProperty.__get__
>     val = self.func(obj)
> pandas/core/groupby/ops.py:299: in group_info
>     comp_ids, obs_group_ids = self._get_compressed_labels()
> pandas/core/groupby/ops.py:315: in _get_compressed_labels
>     all_labels = [ping.labels for ping in self.groupings]
> pandas/core/groupby/ops.py:315: in <listcomp>
>     all_labels = [ping.labels for ping in self.groupings]
> pandas/core/groupby/grouper.py:405: in labels
>     self._make_labels()
> pandas/core/groupby/grouper.py:424: in _make_labels
>     labels = self.grouper.label_info
> pandas/_libs/properties.pyx:34: in pandas._libs.properties.CachedProperty.__get__
>     val = self.func(obj)
> pandas/core/groupby/ops.py:310: in label_info
>     sorter = np.lexsort((labels, self.indexer))
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> args = ((array([ 0, 29]), array([2, 1, 0])),), kwargs = {}, relevant_args = (array([ 0, 29]), array([2, 1, 0]))
> 
> >   ???
> E   ValueError: all keys need to be the same shape
> 
> <__array_function__ internals>:6: ValueError
> ============================================================ 1 failed in 0.17s =============================================================
> ```

@alichaudry that PR is not merged, it *may* or may not pass the test you created; its not clear if this is a different issue or not. happy for you to investigate.

**Comment by alichaudry** [comment]
> @alichaudry that PR is not merged, it _may_ or may not pass the test you created; its not clear if this is a different issue or not. happy for you to investigate.

So what I meant is I took the following steps:

1. cloned #29131 locally (the PR you mentioned)
2. built pandas from scratch in a separate conda env (given that PR author's code)
3. put my test into the local version of #29131 and ran it, which failed

My point is, from what I can tell [my test](https://github.com/pandas-dev/pandas/pull/29800/files#diff-6f020a6c8d0f692c7f6e87cfe0775fe7R280) doesn't work on the code/changes in #29131, which means those changes don't address the bug that I found.

**Comment by alichaudry** [comment]
> @alichaudry that PR is not merged, it _may_ or may not pass the test you created; its not clear if this is a different issue or not. happy for you to investigate.

@jreback So what I meant is I took the following steps:

1. cloned #29131 locally (the PR you mentioned)
2. built pandas from scratch in a separate conda env (given that PR author's code)
3. put my test into the local version of #29131 and ran it, which failed

My point is, from what I can tell [my test](https://github.com/pandas-dev/pandas/pull/29800/files#diff-6f020a6c8d0f692c7f6e87cfe0775fe7R280) doesn't work on the code/changes in #29131, which means those changes don't address the bug that I found.

**Comment by jreback** [comment]
@alichaudry well i suspect that that PR solves some of the problem

happy to take a partial or full patch

File: pandas/core/groupby/grouper.py [hunk 4]
```diff
@@ -155,6 +162,8 @@ def _set_grouper(self, obj: FrameOrSeries, sort: bool = False):
             whether the resulting grouper should be sorted
         """
         assert obj is not None
+        if not obj.equals(self.obj):
```
Comment (jbrockmendel) [inline pandas/core/groupby/grouper.py]: does `obj.equals(self.obj)` vs `obj is self.obj` make a difference?

Comment (alichaudry) [reply pandas/core/groupby/grouper.py]: It shouldn't -- I'll update it

**Comment by jreback** [comment]
@alichaudry pls merge master

**Comment by WillAyd** [comment]
@alichaudry looks like CI is red - can you investigate and try to get passing?

**Comment by alichaudry** [comment]
@WillAyd At this point I'm not sure how to proceed with this PR. I know what the issue is, and I've written a test in this PR which has the correct pass/fail behavior. I also put in a hacky solution for `pd.Grouper` in this PR which passes my test but fails a number of other tests. I don't have the bandwidth right now to re-work the `pd.Grouper` class to fix the root cause so I'm not sure where to go with this.

This issue as I described it in #26564 comes from the fact that I use `pd.Grouper` in a stateless way -- but when I use the grouper in a group-by, say `df_first.groupby(my_grouper)`, and then use the same grouper in another group-by: `df_second.groupby(my_grouper)` it fails as `my_grouper` was mutated by the first group-by and the internal state of the grouper object (and its attributes) makes it fail on the latter one.

**Comment by jbrockmendel** [comment]
@alichaudry Will may have some more specific thoughts, but worst-case scenario you can make a PR with just the test and xfail it, so we'll know if this gets fixed down the road.

**Comment by alichaudry** [comment]
@jbrockmendel followed your advice; updated the PR to include only the test and marked it as xfail. I am seeing some unrelated test errors (they seem to have to do with pyarrow/feather functionality) so I was wondering if you have any insight into this. I will still wait for feedback from @WillAyd.

**Comment by jbrockmendel** [comment]
The pyarrow/feather failures are unrelated, you can ignore them for now.

**Comment by alichaudry** [comment]
@WillAyd If you get a chance I'd love to hear your thoughts on this PR. It's passed all tests as expected (only failing unrelated pyarrow/feather tests).

File: pandas/tests/resample/test_resampler_grouper.py [hunk 5]
```diff
@@ -295,3 +296,35 @@ def test_median_duplicate_columns():
     result = df.resample("5s").median()
     expected.columns = result.columns
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.xfail(reason="marked as xfail for: #26564")
```
Comment (jbrockmendel) [inline pandas/tests/resample/test_resampler_grouper.py]: giving a GH reference is good, but the rest of this doesnt give much information.  is there a single-line description that a reader would find informative?

can you add a "GH" in front of the "#" pls

ca

**Comment by jbrockmendel** [comment]
The OP says that this closes #26564, is that still accurate?

**Comment by WillAyd** [comment]
@alichaudry if you can merge master to get CI green can take another look

**Comment by jreback** [comment]
closing as stale if you want to continue, please open a new PR.

**Pull request closed without merging**